pkg/util/wsstream/conn.go

```diff
@@ -0,0 +1,306 @@
+				config.Protocol = []string{protocol}
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("requested protocol(s) are not supported: %v; supports %v", config.Protocol, allowed)
+}
+
+// Conn supports sending multiple binary channels over a websocket connection.
+// Supports only the "channel.k8s.io" subprotocol.
+type Conn struct {
+	channels []*websocketChannel
+	codec    codecType
+	ready    chan struct{}
+	ws       *websocket.Conn
+	timeout  time.Duration
+}
+
+// NewConn creates a WebSocket connection that supports a set of channels. Channels begin each
+// web socket message with a single byte indicating the channel number (0-N). 255 is reserved for
+// future use. The channel types for each channel are passed as an array, supporting the different
+// duplex modes. Read and Write refer to whether the channel can be used as a Reader or Writer.
+func NewConn(channels ...ChannelType) *Conn {
+	conn := &Conn{
+		ready:    make(chan struct{}),
+		channels: make([]*websocketChannel, len(channels)),
+	}
+	for i := range conn.channels {
+		switch channels[i] {
+		case ReadChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), true, false)
+		case WriteChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), false, true)
+		case ReadWriteChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), true, true)
+		case IgnoreChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), false, false)
+		}
+	}
+	return conn
+}
+
+// SetIdleTimeout sets the interval for both reads and writes before timeout. If not specified,
+// there is no timeout on the connection.
+func (conn *Conn) SetIdleTimeout(duration time.Duration) {
+	conn.timeout = duration
+}
+
+// Open the connection and create channels for reading and writing.
+func (conn *Conn) Open(w http.ResponseWriter, req *http.Request) ([]io.ReadWriteCloser, error) {
+	go func() {
+		defer util.HandleCrash()
+		defer conn.Close()
+		websocket.Server{Handshake: conn.handshake, Handler: conn.handle}.ServeHTTP(w, req)
+	}()
+	<-conn.ready
+	rwc := make([]io.ReadWriteCloser, len(conn.channels))
+	for i := range conn.channels {
+		rwc[i] = conn.channels[i]
+	}
+	return rwc, nil
+}
+
+func (conn *Conn) initialize(ws *websocket.Conn) {
+	protocols := ws.Config().Protocol
+	switch {
+	case len(protocols) == 0, protocols[0] == channelWebSocketProtocol:
+		conn.codec = rawCodec
+	case protocols[0] == base64ChannelWebSocketProtocol:
+		conn.codec = base64Codec
+	}
+	conn.ws = ws
+	close(conn.ready)
+}
+
+func (conn *Conn) handshake(config *websocket.Config, req *http.Request) error {
+	return handshake(config, req, []string{channelWebSocketProtocol, base64ChannelWebSocketProtocol})
+}
+
+func (conn *Conn) resetTimeout() {
+	if conn.timeout > 0 {
+		conn.ws.SetDeadline(time.Now().Add(conn.timeout))
+	}
+}
+
+// Close is only valid after Open has been called
+func (conn *Conn) Close() error {
+	<-conn.ready
+	conn.ws.Close()
+	for _, s := range conn.channels {
```

Comment:
I will change the order, if you have pending data behavior is undefined anyway (you should finish your writing before closing).